File: linear_regression.jl
#!/usr/bin/julia

function get_linear_reg_function(X, Y)
  #println(X)
  #println(Y)
  #println(X')
  #println(X' * X)
  #println(inv(X' * X))
  #println(X' * Y)

  θs = inv(X' * X) * X' * Y

  println("θs")
  println(typeof(θs))
  println(θs)

  return x -> sum(cross(x, θs))
end


# error = 1/2 Σ( | hx - yx |^2 )
function least_squares_error(regression, features::Array{Float64,2}, truth::Array{Float64,1})
  totalerror = 0
  for i = 1:size(truth, 1)
    item = features'[:,i]
    println("item")
    println(typeof(item))
    println(item)
    println("truth")
    println(truth[i,:])
    println("guess")
    println(regression(item))
    error = abs(truth[i,:] - regression(item))^2
    totalerror = totalerror + error
  end
  return totalerror
end




#testing
function test_toy()
  # http://easycalculation.com/statistics/learn-regression.php
  data = [1.0 1.0 2 3 4 5; 2.0 2 10 4 5 6]
  features = data[:,2:]
  truth = data[:,1]

  reg = get_linear_reg_function(features, truth)
  error = least_squares_error(reg, features, truth)

  println(reg)
  println(error)
end

data_dir = "../../data/HW1/"
function test_spam()
  # spam
  spam_data_file = string(data_dir, "spambase/spambase.data")

  println("reading in $spam_data_file")
  spam_data = readcsv(spam_data_file)
end

function test_housing()
  housing_train_data_file = string(data_dir, "housing/housing_train.txt")
  housing_test_data_file = string(data_dir, "housing/housing_test.txt")

  println("reading in $housing_train_data_file")
  housing_train_data = readcsv(housing_train_data_file)
  housing_test_data = readcsv(housing_test_data_file)

  reg = get_linear_reg_function(housing_train_data[:,2:], housing_train_data[:,1])
  error = least_squares_error(reg, housing_test_data[:,2:], housing_test_data[:,1])
  println(error)
end

#main
#test_toy()
test_housing()

File: linear_regression.py
#!/usr/bin/python
import numpy as np
from pprint import pprint
import inspect
import csv

def get_linear_reg_function(X, Y):
  thetas = (X.T * X).I * X.T * Y
  #pprint("thetas")
  #pprint(thetas.A1)
  return lambda x: np.dot(x, thetas.A1)

def least_squares_error(regression, features, truths):
  error = 0

  for i in range(0,len(truths)):
    item = features[i,:].A1
    truth = truths.A1[i]

    #pprint((truth, regression(item)))

    error += pow(abs(truth - regression(item)), 2)

  return error



def read_csv_as_numpy_matrix(filename):
  return np.matrix(list(csv.reader(open(filename,"rb"),delimiter=','))).astype('float')

def mean_squared_error(guesses, truths):
  error = 0
  for guess, truth in zip(guesses, truths):
    error += pow(abs(guess - truth),2)
  return error / len(truths)


def guess_all(regression, features):
  guesses = []
  for item in features:
    guesses.append(regression(item.A1))
  return guesses




import unittest

data_dir = "../../data/HW1/"
class TestLinearReg(unittest.TestCase):


  def test_simple_case(self):
    data = np.matrix('3 3 -2; 5 0 3; 4 4 4')
    regression = get_linear_reg_function(data[:,1:], data[:,0])
    error = least_squares_error(regression, data[:,1:], data[:,0])

  def test_housing(self):
    housing_train_filename = data_dir + "housing/housing_train.txt"
    housing_test_filename = data_dir + "housing/housing_test.txt"

    train_data = read_csv_as_numpy_matrix(housing_train_filename)
    test_data = read_csv_as_numpy_matrix(housing_test_filename)

    features = train_data[:,:12]
    truth = train_data[:,13]
    regression = get_linear_reg_function(features, truth)

    features = test_data[:,:12]
    truth = test_data[:,13]
    error = least_squares_error(regression, features, truth)

    pprint("MSE housing")
    pprint(error / truth.size)

    self.assertTrue(False)

  def test_spam(self):
    spam_filename = data_dir + "spambase/spambase.data"
    data = read_csv_as_numpy_matrix(spam_filename)

    features = data[:,:56]
    truth = data[:,57]
    regression = get_linear_reg_function(features, truth)

    error = least_squares_error(regression, features, truth)

    pprint("MSE spam")
    pprint(error / truth.size)

    self.assertTrue(False)
